### PR TITLE
Add RBS for kaminari-core

### PR DESCRIPTION
**Pull request by m11o**
This PR added RBS files and tests for often used methods in kaminari-core.
I'm going to add [kaminari-activerecord](https://github.com/kaminari/kaminari/tree/00657f252244413b512349f4de20c63ee478baa5/kaminari-activerecord) and [kaminari-actionview](https://github.com/kaminari/kaminari/tree/00657f252244413b512349f4de20c63ee478baa5/kaminari-actionview) gem in another PRs.
Please check it.

[kaminari/kaminari-core](https://github.com/kaminari/kaminari/tree/00657f252244413b512349f4de20c63ee478baa5/kaminari-core)

### CODE REVIEW

File: gems/kaminari-core/1.2/_test/test.rbs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (pocke) [inline gems/kaminari-core/1.2/_test/test.rbs:1]: This line looks unnecessary

File: gems/kaminari-core/1.2/kaminari-core.rbs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    public attr_accessor max_per_page: Integer?
+    public attr_accessor window: Integer
+    public attr_accessor outer_window: Integer
+    public attr_accessor left: Integer
+    public attr_accessor right: Integer
+    public attr_accessor page_method_name: Symbol | String | :page
+    public attr_accessor max_pages: Integer?
+    public attr_accessor params_on_first_page: bool
+
+    public attr_writer param_name: Symbol | String | :page | untyped
```
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:16]: `attr_*`s generate public methods by default, so the `public` notation is not necessary.

File: gems/kaminari-core/1.2/kaminari-core.rbs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  def self.configure: () { (Kaminari::Config) -> void } -> void
+
+  class Config
+    public attr_accessor default_per_page: Integer
+    public attr_accessor max_per_page: Integer?
+    public attr_accessor window: Integer
+    public attr_accessor outer_window: Integer
+    public attr_accessor left: Integer
+    public attr_accessor right: Integer
+    public attr_accessor page_method_name: Symbol | String | :page
```
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:12]: `Symbol` includes `:page`, so `| :page` looks unnecessary.

File: gems/kaminari-core/1.2/kaminari-core.rbs (hunk 2 above)
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:16]: `param_name` attribute accepts an object that receives `call` method.
How about defining an interface?

File: gems/kaminari-core/1.2/kaminari-core.rbs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public attr_accessor max_pages: Integer?
+    public attr_accessor params_on_first_page: bool
+
+    public attr_writer param_name: Symbol | String | :page | untyped
+
+    def param_name: () -> (Symbol | String | :page)
+  end
+
+  module PageScopeMethods
+    include Kaminari::_Model_Relation
```
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:22]: `_Model_Relation` provides `limit` and `offset` methods, but actually `PageScopeMethods` doesn't define them. These methods are defined by a module that includes `PageScopeMethods`.


In this case, "self type parameter" feature is useful.
https://github.com/ruby/rbs/blob/6c847d1177e6e03435cd7214edf85a61e4056d37/docs/syntax.md#module-declaration
> Module declaration takes optional self type parameter, which defines a constraint about a class when the module is mixed.



How about using it?

File: gems/kaminari-core/1.2/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# manifest.yaml describes dependencies which do not appear in the gemspec.
+# If this gem includes such dependencies, comment-out the following lines and
+# declare the dependencies.
+# If all dependencies appear in the gemspec, you can remove this file.
```
Comment (pocke) [inline gems/kaminari-core/1.2/manifest.yaml:4]: Could you remove this file?

**Review by pocke — CHANGES_REQUESTED** [review]
Thanks for your PR! I left some comments, please check them 👍

File: gems/kaminari-core/1.2/kaminari-core.rbs (hunk 4 above)
Comment (m11o) [reply gems/kaminari-core/1.2/kaminari-core.rbs]: Thanks! It was difficult to write the expression.
I tried it. Please check it. e79ad94412f8fe27a5d2c9a236657f14fc3f1e79

File: gems/kaminari-core/1.2/kaminari-core.rbs (hunk 2 above)
Comment (m11o) [reply gems/kaminari-core/1.2/kaminari-core.rbs]: Great idea! Thank you! I fixed it. 844cb848f7345fab7bf32ba298ccb78afc717acc

File: gems/kaminari-core/1.2/kaminari-core.rbs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    attr_accessor page_method_name: Symbol | String
+    attr_accessor max_pages: Integer?
+    attr_accessor params_on_first_page: bool
+
+    attr_writer param_name: Symbol | String | _Caller
+
+    def param_name: () -> (Symbol | String)
+  end
+
+  module PageScopeMethods[Relation] : _Model_Relation[Relation]
```
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:21]: Is `Relation` type parameter necessary?
`self` is appropriate for the returned type of `per`, `limit` and `offset` methods.

File: gems/kaminari-core/1.2/kaminari-core.rbs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    def last_page?: () -> bool
+    def out_of_range?: () -> bool
+  end
+
+  interface _Model_Relation[Relation]
+    def limit: (Integer) -> Relation
+    def offset: (Integer) -> Relation
+  end
+
+  interface _Caller
```
Comment (pocke) [inline gems/kaminari-core/1.2/kaminari-core.rbs:37]: IMO: `_Caller` naming feels a little bit strange. In programming terms, caller means method caller.
How about `_Callable` instead? `-able` is the major pattern for interface naming.

**Comment by pocke** [comment]
The CI fails due to many commits. This problem has been fixed by #223 probably. So could you merge the main branch into this branch?

**Review by pocke — APPROVED** [review]
Thanks!

**Pull request merged**